src/Tutor.Core/LearnerModel/DomainOverlay/KnowledgeComponentMasteries/KnowledgeComponentMastery.cs

```diff
@@ -70,40 +68,29 @@ private void JoinOrLaunchSession()
         if (!SessionTracker.HasUnfinishedSession) LaunchSession();
     }
 
-    public Result<List<InstructionalItem>> SelectInstructionalItems()
+    public Result RecordInstructionalItemSelection()
     {
         JoinOrLaunchSession();
-
         Causes(new InstructionalItemsSelected());
-        return Result.Ok(KnowledgeComponent.InstructionalItems.OrderBy(i => i.Order).ToList());
+        return Result.Ok();
     }
 
-    public Result<int> SelectAssessmentItem(IAssessmentItemSelector assessmentItemSelector)
+    public Result RecordAssessmentItemInteraction(AssessmentItemEvent interaction)
     {
-        var result = assessmentItemSelector.SelectSuitableAssessmentItemId(AssessmentItemMasteries, IsPassed);
-        if (result.IsFailed) return result;
-
-        var aim = AssessmentItemMasteries.Find(aim => aim.AssessmentItemId == result.Value);
-        if (aim == null) return NoAssessmentItemWithId(result.Value);
+        var aim = AssessmentItemMasteries.Find(aim => aim.AssessmentItemId == interaction.AssessmentItemId);
+        if (aim == null) return Result.Fail("No mastery for assessment item with id " + interaction.AssessmentItemId +
+                                            ". Were masteries created and loaded correctly?");
 
         JoinOrLaunchSession();
-        aim.Select();
+        var result = aim.RecordInteraction(interaction);
+        if (result.IsSuccess)
+        {
+            if (interaction is IPassCausing) TryPass();
```

Comment:
I don't think that introducing interfaces in this way is a good idea. Essentially, you are using interfaces as a constant property, which is uncommon and therefore confusing. I think that a better option would be to define unserializable boolean properties in the base class that are false and redefine them in the appropriate children.

However, I am also not sure if this is a good idea. Maybe remove it all together and call TryPass and TryComplete all the time?

Reply:
Always calling <code>TryPass</code> and <code>TryComplete</code> makes a lot of sense to me theoretically, which is why it was my initial approach as well. I removed the interfaces in 637dbafa2b0736a8b80ca0fe3ade7270792b82d9. I don't actually think it should be an issue in practice either, since calculations performed in these methods aren't that complicated, so trying to mark events which might cause pass or completion is probably just more trouble than it's worth.